File: problems/0095-unique-binary-search-trees-ii/solution.py
class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        """
        Problem: 0095 - Unique Binary Search Trees II

        Given an integer n, return all structurally unique BSTs (binary search trees)
        that store values 1 to n.

        Approach:
        - Use recursion with manual memoization to avoid redundant subtree construction.
        - For each (start, end) interval, store already computed trees in a memo table.
        - For each root value i in [start, end]:
            - Recursively build all left subtrees in [start, i - 1]
            - Recursively build all right subtrees in [i + 1, end]
            - Combine each left and right subtree with root i
        - Return all valid BSTs rooted in this interval.

        Time complexity: O(C_n), where C_n is the nth Catalan number ≈ O(4^n / n^1.5)
        Space complexity: O(C_n) for result trees, O(n^2) for memo table, O(n) stack depth
        """
        if n == 0:
            return []

        memo = {}  # Key: (start, end) → Value: list of TreeNode roots

        def build(start: int, end: int) -> List[Optional[TreeNode]]:
            # Return memoized result if already computed.
            if (start, end) in memo:
                return memo[(start, end)]

            # Base case: empty subtree.
            if start > end:
                return [None]

            all_trees = []

            # Try every number i in range [start, end] as root.
            for i in range(start, end + 1):
                # Recursively build all left subtrees from [start, i-1].
                left_subtrees = build(start, i - 1)

                # Recursively build all right subtrees from [i+1, end].
                right_subtrees = build(i + 1, end)

                # Combine each left tree and right tree with root i.
                for left in left_subtrees:
                    for right in right_subtrees:
                        root = TreeNode(i)
                        root.left = left
                        root.right = right
                        all_trees.append(root)

            # Memoize result for current (start, end).
            memo[(start, end)] = all_trees
            return all_trees

        return build(1, n)

File: problems/0095-unique-binary-search-trees-ii/solution.cpp
class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        /*
        Problem: 0095 - Unique Binary Search Trees II

        Given an integer n, return all structurally unique BSTs (binary search trees)
        that store values 1 to n.

        Approach:
        - Use recursion with manual memoization to avoid recomputing subtrees.
        - For each interval (start, end), cache the result list of BST roots.
        - For every possible root value i in [start, end]:
            - Recursively build all left subtrees from [start, i - 1].
            - Recursively build all right subtrees from [i + 1, end].
            - Combine each left and right with a new root node i.
        - Return the cached result if already computed.

        Time complexity: O(C_n), where C_n is the nth Catalan number ≈ O(4^n / n^1.5)
        Space complexity: O(C_n) for result trees, O(n^2) memoization, O(n) recursion stack
        */
        if (n == 0) {
            return {};
        }

        return build(1, n);
    }

private:
    // Memoization table: key = (start, end), value = list of trees for that interval.
    map<pair<int, int>, vector<TreeNode*>> memo;

    // Recursively build all unique BSTs for values in the range [start, end].
    vector<TreeNode*> build(int start, int end) {
        // Return memoized result if available.
        if (memo.count({start, end})) {
            return memo[{start, end}];
        }

        // Base case: empty subtree → return vector with one nullptr.
        if (start > end) {
            return {nullptr};
        }

        vector<TreeNode*> all_trees;

        // Try each number in the range as the root.
        for (int i = start; i <= end; ++i) {
            // Recursively generate all left subtrees.
            vector<TreeNode*> left_subtrees = build(start, i - 1);

            // Recursively generate all right subtrees.
            vector<TreeNode*> right_subtrees = build(i + 1, end);

            // Combine left and right with root i.
            for (TreeNode* left : left_subtrees) {
                for (TreeNode* right : right_subtrees) {
                    TreeNode* root = new TreeNode(i);
                    root->left = left;
                    root->right = right;
                    all_trees.push_back(root);
                }
            }
        }

        // Cache and return the result for (start, end).
        memo[{start, end}] = all_trees;
        return all_trees;
    }
};

